spec/services/atmosphere/cloud/satisfy_appliance_spec.rb

```diff
@@ -17,6 +17,95 @@
   let!(:openstack) { create(:openstack_with_flavors, funds: [fund]) }
   let!(:tmpl_of_shareable_at) { create(:virtual_machine_template, appliance_type: shareable_appl_type, tenants: [openstack])}
 
+  context '#assign fund' do
+    let!(:t) { create(:openstack_with_flavors, active: true, funds: [fund]) }
+    let!(:vmt) { create(:virtual_machine_template, tenants: [t]) }
+    let!(:appliance_type) do
+      create(
+        :appliance_type,
+        preference_cpu: 0,
+        preference_disk: 0,
+        preference_memory: 0,
+        virtual_machine_templates: [vmt]
+      )
+    end
+    let!(:appliance_set) { create(:appliance_set, user: u) }
+    let!(:appliance) do
+      create(:appliance,
+             appliance_set: appliance_set,
+             appliance_type: appliance_type,
+             fund: nil)
+    end
+
+    it 'does not change fund when externally assigned' do
+      funded_appliance = create(:appliance)
+      expect(funded_appliance.fund).not_to eq appliance.send(:default_fund)
+    end
+
+    it 'gets default fund from its user if no fund is set' do
+      appliance.reload
+      Atmosphere::Cloud::SatisfyAppliance.new(appliance).execute
+      expect(appliance.fund).to eq appliance.send(:default_fund)
+    end
+
+    it 'prefers default fund if it supports relevant tenant' do
+      appliance.reload
+      default_t = create(
+        :openstack_with_flavors,
+        active: true,
+        funds: [appliance_set.user.default_fund]
+      )
+      funded_t = create(
+        :openstack_with_flavors,
+        active: true,
+        funds: [create(:fund)]
+      )
+      appliance_set.user.funds << funded_t.funds.first
+      create(:virtual_machine_template,
+             appliance_type: appliance_type,
+             tenants: [default_t])
+      create(:virtual_machine_template,
+             appliance_type: appliance_type,
+             tenants: [funded_t])
+
+      Atmosphere::Cloud::SatisfyAppliance.new(appliance).execute
+
+      expect(appliance.fund.reload).not_to eq funded_t.funds.first.reload
```

Comment:
You can assign fund created in line 61 into variable, then you will not need to traverse over relations.